### PR TITLE
Nktchinh

### PR DESCRIPTION
**Pull request by ntkchinh**

### CODE REVIEW

File: assets/js/linear_algebra/dot_product_formula.js [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+      y: v.y * uTv,
+      z: v.z * uTv,
+      r: 1.8,
+      color: 'grey'
+  }
+
+  let uTvv_line = [
+      {x: 0, y: 0, z: 0},
+      {x: uTvv.x, y: uTvv.y, z: uTvv.z,
+       color: 0, tt: true}
```
Comment (thtrieu) [inline assets/js/linear_algebra/dot_product_formula.js]: set uTvv_line.color = 0 instead

File: assets/js/linear_algebra/dot_product_formula.js [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+      texts_to_show, 
+      start_coord_x=-2.2, start_coord_y=2.2,
+      col_unit=0.24, row_unit=0.3,
+      cols_width_array=[1.5, 1.2, 2.2, 0.8, 2.2, 0.8, 2.2, 0.8, 2.2, 0.8, 2.2, 0.8, 2.2],
+      rows_width_array=[1.0, 1.8, 1.0])
+  );
+}
+
+function init(){
+  scatter = [];
```
Comment (thtrieu) [inline assets/js/linear_algebra/dot_product_formula.js]: there is no need for this line.

File: assets/js/linear_algebra/dot_product_formula.js [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+
+
+return {
+  init: function(){init();}
+};
+
+})();
+
+
+
```
Comment (thtrieu) [inline assets/js/linear_algebra/dot_product_formula.js:247]: rm

File: assets/js/linear_algebra/dot_product_formula2d.js [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+  scale = 60, 
+  scatter = [], 
+  axis = [],
+  expectedAxis = [],
+  startAngleX = Math.PI,
+  startAngleY = 0.,
+  startAngleZ = 0.;
+let  axis_len = 2;
+
+let unit = axis_len/10;
```
Comment (thtrieu) [inline assets/js/linear_algebra/dot_product_formula2d.js]: merge these two lines into the same "let"?

File: assets/js/linear_algebra/dot_product_formula2d.js [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+  lib.plot_texts(lib.text_table_to_list(
+      texts_to_show,
+      start_coord_x=-2.2, start_coord_y=2.2,
+      col_unit=0.26, row_unit=0.3,
+      cols_width_array=[1.8, 1.2, 2, 0.8, 2, 0.8, 2, 0.8, 2],
+      rows_width_array=[1.0, 1.8, 1.0])
+  );
+
+  lib.sort();
+
```
Comment (thtrieu) [inline assets/js/linear_algebra/dot_product_formula2d.js]: rm

File: assets/js/linear_algebra/dot_product_formula2d.js [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+  expectedScatter = lib.rotate_points(scatter, startAngleX, startAngleY, startAngleZ);
+  expectedAxis = lib.rotate_lines(axis, startAngleX, startAngleY, startAngleZ);
+  plot(expectedScatter, 
+       expectedAxis, 
+       1000);
+  drag_end();
+
+  text = [];
+
```
Comment (thtrieu) [inline assets/js/linear_algebra/dot_product_formula2d.js]: rm

File: assets/js/linear_algebra/dot_product_formula2d.js [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+  scatter = [u, v];
+
+  expectedScatter = lib.rotate_points(scatter, startAngleX, startAngleY, startAngleZ);
+  expectedAxis = lib.rotate_lines(axis, startAngleX, startAngleY, startAngleZ);
+  plot(expectedScatter, 
+       expectedAxis, 
+       1000);
+  drag_end();
+
+  text = [];
```
Comment (thtrieu) [inline assets/js/linear_algebra/dot_product_formula2d.js]: rm

Comment (thtrieu) [reply assets/js/linear_algebra/dot_product_formula2d.js]: what is this "text" variable?

File: assets/js/linear_algebra/lib.js [hunk 8]
```diff
@@ -618,45 +618,42 @@ function create_segments(d, k=10) {
...
-                           col_unit * cols_width_array[j-1]);
-    };
-
-    // setting the rows' coords:
-    row_coords.push(start_coord_y);
-    for (var i = 1; i < numb_of_rows; i++) {
-        row_coords.push(row_coords[i-1] +
-                           row_unit * rows_width_array[i-1]);
-    }
+                            cols_width_array, rows_width_array
```
Comment (thtrieu) [inline assets/js/linear_algebra/lib.js]: better names: `col_heights` and `row_widths`

File: assets/js/linear_algebra/dot_product_formula2d.js [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+                      dragged_point(i);
+                    } else {
+                      dragged_point_only();
+                    }
+                  },
+                  drag_start_fn=drag_start,
+                  drag_end_fn=drag_end);
+
+  // setting up the texts need to show;
+
```
Comment (thtrieu) [inline assets/js/linear_algebra/dot_product_formula2d.js]: rm

File: assets/js/linear_algebra/lib.js [hunk 10]
```diff
@@ -618,45 +618,42 @@ function create_segments(d, k=10) {
...
-    for (var i = 0; i < numb_of_rows; i++) {
-      for (var j = 0; j < numb_of_cols; j++) {
-        text_to_plot = texts[i][j];
-        text_to_plot.x = col_coords[j];
-        text_to_plot.y = row_coords[i];
-        text_to_plot.text_opacity = 1;
-        list_of_text.push(text_to_plot);
-      };
-    }
+  var list_of_text = [];
```
Comment (thtrieu) [inline assets/js/linear_algebra/lib.js]: `list_of_texts`

**Review by thtrieu — CHANGES_REQUESTED** [review]

File: assets/js/linear_algebra/dot_product_formula2d.js (hunk 7 above)
Comment (ntkchinh) [reply assets/js/linear_algebra/dot_product_formula2d.js]: done

File: assets/js/linear_algebra/lib.js (hunk 8 above)
Comment (thtrieu) [reply assets/js/linear_algebra/lib.js]: or shorter names: `dhs` and `dws` (also sound more mathy)

**Review by thtrieu — CHANGES_REQUESTED** [review]

File: assets/js/linear_algebra/dot_product_formula.js (hunk 1 above)
Comment (ntkchinh) [reply assets/js/linear_algebra/dot_product_formula.js]: done

File: assets/js/linear_algebra/dot_product_formula.js (hunk 2 above)
Comment (ntkchinh) [reply assets/js/linear_algebra/dot_product_formula.js]: done

File: assets/js/linear_algebra/dot_product_formula.js (hunk 3 above)
Comment (ntkchinh) [reply assets/js/linear_algebra/dot_product_formula.js:247]: done

File: assets/js/linear_algebra/dot_product_formula2d.js (hunk 4 above)
Comment (ntkchinh) [reply assets/js/linear_algebra/dot_product_formula2d.js]: done

File: assets/js/linear_algebra/dot_product_formula2d.js (hunk 5 above)
Comment (ntkchinh) [reply assets/js/linear_algebra/dot_product_formula2d.js]: done

File: assets/js/linear_algebra/dot_product_formula2d.js (hunk 6 above)
Comment (ntkchinh) [reply assets/js/linear_algebra/dot_product_formula2d.js]: done

File: assets/js/linear_algebra/lib.js (hunk 8 above)
Comment (ntkchinh) [reply assets/js/linear_algebra/lib.js]: done

**Pull request merged**